src/main/java/io/divolte/server/ValidatedConfiguration.java

```diff
@@ -0,0 +1,535 @@
+            exceptions.add(ce);
+            return Optional.empty();
+        }
+    }
+
+    private static final Joiner COMMA_JOINER = Joiner.on(',');
+
+    private static Properties configToProperties(final Config config, final String path) {
+        final Properties properties = new Properties();
+        for (final Map.Entry<String,ConfigValue> entry : config.getConfig(path).entrySet()) {
+            final ConfigValue configValue = entry.getValue();
+            final String value;
+            switch (configValue.valueType()) {
+                case STRING:
+                case BOOLEAN:
+                case NUMBER:
+                    value = configValue.unwrapped().toString();
+                    break;
+                case LIST:
+                    final ConfigList configList = (ConfigList)configValue;
+                    // We only need to support 'simple' types here.
+                    value = COMMA_JOINER.join(configList.unwrapped());
+                    break;
+                case OBJECT:
+                case NULL:
+                default:
+                    throw new IllegalStateException("Property type not supported for Kafka configuration: " + entry);
+            }
+            properties.setProperty(entry.getKey(), value);
+        }
+        return properties;
+    }
+
+    /**
+	 * Returns the validated configuration object tree. This is only returned
+	 * when no validation errors exist. The method throws
+	 * {@code IllegalStateException} otherwise.
+	 * 
+	 * @return The validated configuration.
+	 * @throws IllegalStateException
+	 *             When validation errors exist.
+	 */
+    public DivolteConfiguration configuration() {
+        if (exceptions.size() > 0) {
+            throw new IllegalStateException("Attempt to access invalid configuration.");
+        }
+        return divolteConfiguration;
+    }
+
+    /**
+	 * Returns a list of {@code ConfigException} that were thrown during
+	 * configuration validation.
+	 * 
+	 * @return A list of {@code ConfigException} that were thrown during
+	 *         configuration validation.
+	 */
+    public List<ConfigException> errors() {
+        return exceptions;
+    }
+
+    /**
+	 * Returns false if validation errors exist, true otherwise.
+	 * 
+	 * @return false if validation errors exist, true otherwise.
+	 */
+    public boolean isValid() {
+        return exceptions.size() == 0;
+    }
+
+    @ParametersAreNullableByDefault
+    public final static class DivolteConfiguration {
+        public final ServerConfiguration server;
+        public final TrackingConfiguration tracking;
+        public final JavascriptConfiguration javascript;
+        public final IncomingRequestProcessorConfiguration incomingRequestProcessor;
+        public final KafkaFlusherConfiguration kafkaFlusher;
+        public final HdfsFlusherConfiguration hdfsFlusher;
+
+        public DivolteConfiguration(
```

Comment:
I think this constructor should be private, or we need to guard against null parameters.